src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -171,11 +172,56 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     }
 
                     break;
+                case ODataReaderState.DeltaResourceSetStart:
+                    ODataDeltaResourceSet deltaResourceSet = (ODataDeltaResourceSet)reader.Item;
+
+                    ODataDeltaResourceSetWrapper deltaResourceSetWrapper = new ODataDeltaResourceSetWrapper(deltaResourceSet);
+
+                    if (itemsStack.Count > 0)
+                    {
+                        AddNestedResourceInfo(itemsStack, deltaResourceSetWrapper);
+                    }
+                    else
+                    {
+                        topLevelItem = deltaResourceSetWrapper;
+                    }
+
+                    itemsStack.Push(deltaResourceSetWrapper);
+                    break;
+
+                case ODataReaderState.DeltaLink:
+                case ODataReaderState.DeltaDeletedLink:
+
+                    //Throw error if Delta Link appears
+                    throw Error.NotSupported(SRResources.DeltaLinkNotSupported);
 
                 default:
                     Contract.Assert(false, "We should never get here, it means the ODataReader reported a wrong state.");
                     break;
             }
         }
+
+
+        private static void AddNestedResourceInfo(Stack<ODataItemBase> itemsStack, ODataResourceSetWrapperBase resourceSetWrapper)
+        {
+            ODataNestedResourceInfoWrapper parentNestedResourceInfo = (ODataNestedResourceInfoWrapper)itemsStack.Peek();
+            Contract.Assert(parentNestedResourceInfo.NestedResourceInfo.IsCollection == true, "Only collection nested properties can contain resource set as their child.");
+            parentNestedResourceInfo.NestedItems.Add(resourceSetWrapper);
+        }
+
+        private static void AddResourceToParent(Stack<ODataItemBase> itemsStack, ODataResourceWrapper resourceWrapper)
+        {
+            ODataItemBase parentItem = itemsStack.Peek();
+            ODataResourceSetWrapperBase parentResourceSet = parentItem as ODataResourceSetWrapperBase;
```

Comment:
Why does this line need to be here and not the `else` block below? The `if` block doesn't seem to depend on this line at all.